Use standard httpd logging format in error log

marcstern opened the pull request:
All credits to @arminabf - see https://github.com/owasp-modsecurity/ModSecurity/pull/1997

Use the server context, like in all other places to use standard httpd format.

Remark: We previously had a mix; some entries were using the standard httpd format, some not.
Entries using the standard httpd format had the [client] field duplicated.
In this PR, we use standard logging for all lines and remove the duplicated field.
Actual changes:
- entries that were not using the standard format will use it
- entries that were using the standard format won't have the [client] field duplicated anymore

Commit:
Use standard httpd logging format in error log

Commit:
Remove redundant entry

[client %s] is added by the standard httpd log function => remove it

fzipi commented:
Is there a way to actually test this change?

theseion commented:
Sounds good, but I'm with @fzipi. From the code it's not clear what the output will be. Does the `[client]` field still contain the same IP address that ModSecurity would have injected in every case?

marcstern commented:
> Sounds good, but I'm with @fzipi. From the code it's not clear what the output will be. Does the `[client]` field still contain the same IP address that ModSecurity would have injected in every case?

Yes, same logic: useragent_ip if present, client_ip otherwise

marcstern commented:
Code from httpd log.c
```
if (info->r) {
    len += apr_snprintf(buf + len, buflen - len,
                        info->r->connection->sbh ? "[client %s:%d] " : "[remote %s:%d] ",
                        info->r->useragent_ip,
                        info->r->useragent_addr ? info->r->useragent_addr->port : 0);
}
else if (info->c) {
    len += apr_snprintf(buf + len, buflen - len,
                        info->c->sbh ? "[client %s:%d] " : "[remote %s:%d] ",
                        info->c->client_ip,
                        info->c->client_addr ? info->c->client_addr->port : 0);
}
```

theseion commented:
That's good. But a test that asserts the log format would be even better.

Commit:
# Send some requests & check log format

Commit:
Merge branch 'v2/pr/errorlog' of https://github.com/marcstern/ModSecurity into v2/pr/errorlog

marcstern commented:
I added some tests

theseion commented:
Thanks, that's great. @airween, isn't there a test suite for ModSecurity where the test could be added?

airween commented on .github/workflows/ci.yml:
```diff
@@ -49,4 +49,14 @@ jobs:
+          # Invalid request
+          curl -s http://127.0.01/%2e%2f > /dev/null || echo $? 
+          # Check log format
+          grep -F ModSecurity < /var/log/apache2/error.log | grep -vP "^\[[^\]]+\] \[security2:[a-z]+\] \[pid [0-9]+:tid [0-9]+\] (?:\[client [0-9.:]+\] )?ModSecurity" || exit 0
```
I'm not sure this test cover the case what you wanted to fix, namely only one `[client]` field can be there - or you still accept the two fields log format?

This pattern:
```
\[pid [0-9]+:tid [0-9]+\] (?:\[client [0-9.:]+\] )
```
matches with this string:
```
[pid 6545:tid 6547] [client ::1:49050] [client ::1]
```
but only till the firs `[client]` field, so this substring:
```
[pid 6545:tid 6547] [client ::1:49050]
```

Was that your intention?

marcstern replied:
`^\[[^\]]+\] \[security2:[a-z]+\] \[pid [0-9]+:tid [0-9]+\] (?:\[client [0-9.:]+\] )?ModSecurity` doesn't match `[pid 6545:tid 6547] [client ::1:49050] [client ::1] ModSecurity ...`

airween replied:
Ah, you're right - that was my mistake.

airween commented:
> Thanks, that's great. @airween, isn't there a test suite for ModSecurity where the test could be added?

Unfortunately no, there isn't.

We have only this CI:

And as @marcstern wrote, he [added](https://github.com/owasp-modsecurity/ModSecurity/pull/3192/files#diff-b803fcb7f17ed9235f1e5cb1fcd2f5d3b2838429d4368ae4c57ce4436577f03fR52-R62) a test (but I'm not sure that's what we expect - I'm waiting for Marc's answer).

But as I introduced few weeks ago, there is [MRTS](https://github.com/owasp-modsecurity/MRTS), where the aim is to create a full covered test suite. Feel free to review and contribute! :smile:

Commit:
Define _FORTIFY_SOURCE=3 & _GLIBCXX_ASSERTIONS that add glibc/libstdc++ assertions.
See https://www.gnu.org/software/libc/manual/html_node/Source-Fortification.html & https://gcc.gnu.org/wiki/LibstdcxxDebugMode

_GLIBCXX_ASSERTIONS is probably useless as we have pure C here, but let's define it in case some checks are included (or will be in a future version).
As we handle some requests here, that may help to trap a problem.
